go/test/endtoend/bootstrap_test.go

```diff
@@ -0,0 +1,549 @@
+		pgctldProcess:  pgctldCmd,
+		multipoolerCmd: multipoolerCmd,
+	}
+}
+
+// waitForMultipoolerReady polls the multipooler gRPC endpoint until it's ready
+func waitForMultipoolerReady(t *testing.T, grpcPort int, timeout time.Duration) {
+	t.Helper()
+
+	deadline := time.Now().Add(timeout)
+	var lastErr error
+	for time.Now().Before(deadline) {
+		conn, err := grpc.NewClient(
+			fmt.Sprintf("localhost:%d", grpcPort),
+			grpc.WithTransportCredentials(insecure.NewCredentials()),
+		)
+		if err != nil {
+			lastErr = err
+			time.Sleep(200 * time.Millisecond)
+			continue
+		}
+
+		client := multipoolermanagerpb.NewMultiPoolerManagerClient(conn)
+		ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
+		_, err = client.InitializationStatus(ctx, &multipoolermanagerdatapb.InitializationStatusRequest{})
+		cancel()
+		conn.Close()
+
+		if err == nil {
+			return // Success!
+		}
+
+		lastErr = err
+		time.Sleep(200 * time.Millisecond)
+	}
+
+	// On failure, log the last error
+	if lastErr != nil {
+		t.Logf("Last error from multipooler: %v", lastErr)
+	}
+
+	// On failure, try to read and display the multipooler log
+	t.Errorf("Multipooler at port %d did not become ready within %v", grpcPort, timeout)
+	t.FailNow()
+}
+
+// cleanupNode stops pgctld and multipooler processes
+func cleanupNode(t *testing.T, node *nodeInstance) {
+	t.Helper()
+	if node.multipoolerCmd != nil && node.multipoolerCmd.Process != nil {
+		_ = node.multipoolerCmd.Process.Kill()
+		_ = node.multipoolerCmd.Wait()
+	}
+	if node.pgctldProcess != nil && node.pgctldProcess.Process != nil {
+		_ = node.pgctldProcess.Process.Kill()
+		_ = node.pgctldProcess.Wait()
+	}
+}
+
+// checkInitializationStatus checks the initialization status of a node
+func checkInitializationStatus(t *testing.T, node *nodeInstance) *multipoolermanagerdatapb.InitializationStatusResponse {
+	t.Helper()
+
+	conn, err := grpc.NewClient(
+		fmt.Sprintf("localhost:%d", node.grpcPort),
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+	)
+	require.NoError(t, err)
+	defer conn.Close()
+
+	client := multipoolermanagerpb.NewMultiPoolerManagerClient(conn)
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+
+	resp, err := client.InitializationStatus(ctx, &multipoolermanagerdatapb.InitializationStatusRequest{})
+	require.NoError(t, err)
+
+	return resp
+}
+
+// connectToPostgres establishes a connection to PostgreSQL using Unix socket
+func connectToPostgres(t *testing.T, socketDir string, port int) *sql.DB {
+	t.Helper()
+
+	connStr := fmt.Sprintf("host=%s port=%d user=postgres dbname=postgres sslmode=disable", socketDir, port)
+	db, err := sql.Open("postgres", connStr)
+	require.NoError(t, err, "Failed to open database connection")
+
+	err = db.Ping()
+	require.NoError(t, err, "Failed to ping database")
+
+	return db
+}
+
+// startEtcd starts an etcd instance for testing
+func startEtcd(t *testing.T, dataDir string) (string, *exec.Cmd) {
```

Comment:
Not blocking: seems like we probably have more versions of `startEtcd()` in the codebase than we need. Wonder if we can consolidate some shared test helpers for that kind of thing?